lib/cn/kcompact.c

```diff
@@ -27,111 +23,18 @@
 #include "cn_tree_compact.h"
 
-/**
- * struct merge_item -- an item in the bin_heap
- */
-struct merge_item {
-    struct key_obj         kobj;
-    struct kvset_iter_vctx vctx;
-    uint                   src;
-};
-
 static int
-merge_item_compare(const void *a_blob, const void *b_blob)
+kv_item_compare(const void *a, const void *b)
 {
-    const struct merge_item *a = a_blob;
-    const struct merge_item *b = b_blob;
-    int                      rc;
+    const struct cn_kv_item *item_a = a;
+    const struct cn_kv_item *item_b = b;
 
-    /* Tie breaker: If keycmp() return 0, the keys are equal, in this case
-     * lower numbered merge sources contain newer data and must come out
-     * of the binheap first.
+    /* In the event the keys are even, the bin heap implementation, will return
+     * the key from earliest element source by index. We can use this as a proxy
+     * for the newest key since the element sources are ordered from newest
+     * kvset to oldest kvset.
      */
-    rc = key_obj_cmp(&a->kobj, &b->kobj);
-    if (rc)
-        return rc;
-    if (a->src < b->src)
-        return -1;
-    if (a->src > b->src)
-        return 1;
-    return 0;
-}
-
-static merr_t
-replenish(struct bin_heap *bh, struct kv_iterator **iterv, uint src, struct cn_merge_stats *stats)
-{
-    struct kv_iterator *iter = iterv[src];
-    merr_t              err;
-    struct merge_item   item;
-
-    if (HSE_UNLIKELY(iter->kvi_eof))
-        return 0;
-
-    err = kvset_iter_next_key(iter, &item.kobj, &item.vctx);
-    if (ev(err))
-        return err;
-    if (HSE_UNLIKELY(iter->kvi_eof))
-        return 0;
-
-    item.src = src;
-
-    err = bin_heap_insert(bh, &item);
-    if (ev(err))
-        return err;
-
-    stats->ms_keys_in++;
-    stats->ms_key_bytes_in += key_obj_len(&item.kobj);
-
-    return 0;
-}
-
-static merr_t
-merge_init(
-    struct bin_heap **     bh_out,
-    struct kv_iterator **  iterv,
-    u32                    iterc,
-    struct cn_merge_stats *stats)
-{
-    u32    i;
-    merr_t err;
-
-    err = bin_heap_create(bh_out, iterc, sizeof(struct merge_item), merge_item_compare);
-    if (ev(err))
-        goto err_exit1;
-
-    stats->ms_srcs = iterc;
-
-    for (i = 0; i < iterc; i++) {
-        err = replenish(*bh_out, iterv, i, stats);
-        if (ev(err))
-            goto err_exit2;
-    }
-
-    return 0;
-
-err_exit2:
-    bin_heap_destroy(*bh_out);
-err_exit1:
-    return err;
-}
-
-/* return true if item returned, false if no more items */
-static HSE_ALWAYS_INLINE bool
-get_next_item(
-    struct bin_heap *      bh,
-    struct kv_iterator **  iterv,
-    struct merge_item *    item,
-    struct cn_merge_stats *stats,
-    merr_t *               err_out)
-{
-    bool got_item;
-
-    got_item = bin_heap_get_delete(bh, item);
-    if (got_item)
-        *err_out = replenish(bh, iterv, item->src, stats);
-    else
-        *err_out = 0;
-    return got_item;
+    return key_obj_cmp(&item_a->kobj, &item_b->kobj);
```

Comment:
Nice comment!  But what does "even" mean?  Equal?

Reply:
Updated the wording to equal and fixed a few grammatical errors for the next push.